projects/aws/image-builder/builder/manifests.go

```diff
@@ -100,3 +117,108 @@ func downloadEKSAManifests(outputPath string) error {
 	}
 	return nil
 }
+
+// extractTarball extracts the provided tarball onto the path
+func extractTarball(tarball, path string) error {
+	tarFile, err := os.Open(tarball)
+	if err != nil {
+		return err
+	}
+	defer tarFile.Close()
+
+	// Create a new directory to extract the tarball into.
+	if err = os.RemoveAll(path); err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(path, 0755)
+	if err != nil {
+		return err
+	}
+
+	// Create a tar reader for the tarball.
+	tarReader := tar.NewReader(tarFile)
+
+	// Iterate over the tarball's entries.
+	for {
+		// Get the next header from the tarball.
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			// We've reached the end of the tarball.
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		// Create a new file for the tar entry.
+		outFile, err := os.Create(filepath.Join(path, header.Name))
+		if err != nil {
+			return err
+		}
+		defer outFile.Close()
+
+		// Copy the tar entry's contents to the new file.
+		_, err = io.Copy(outFile, tarReader)
+		if err != nil {
+			return err
+		}
+
+		outFile.Chmod(os.FileMode(header.Mode))
+	}
+
+	return nil
+}
+
+func extractAndPrepManifestTarball(tarballFile, privateEksDServerDomain, privateEksAServerDomain string) error {
+	log.Println("Manifest tarball provided, extracting to directory")
+	manifestDir, err := getManifestRoot()
+	if err != nil {
+		return fmt.Errorf("Error retrieving manifest root")
+	}
+	if err = extractTarball(tarballFile, manifestDir); err != nil {
+		return fmt.Errorf("Error extracting tarball: %v", err)
+	}
+
+	// Replacing eks-a bundles manifest hostname
+	// These endpoints are used for artifacts like containerd, crictl & etcdadm
+	// Find all eks-a bundles manifest and replace hostname
+	files, err := os.ReadDir(manifestDir)
```

Comment:
Yep none of the files are symlinks. Given theres no sub-directories to recursively walk I used `os.ReadDir`. I don't mind switching to Walk if its consistent across our repos. Both are relatively the same performance in this case.

Reply:
Ah okay, I thought since you're checking for it not being a directory, there would be subdirs to traverse recursively. Since we need just file names, we could also use `Readdirnames` to avoid calling `file.Name()` many times in the loop.
```go
manifestDirFd, err := os.Open(manifestDir)
if err != nil {
    return err
}
defer manifestDirFd.Close()
fileNames, err := manifestDirFd.Readdirnames(0)
if err != nil {
    return err
}
```
This is supposed to be a faster method too but it wouldn't be too noticeable since we only have few files.

I defer to your choice, as it's not a big deal.